### PR TITLE
tree-wide: Introduce free_and_replace_full()

### PR DESCRIPTION
**Pull request by daandemeyer**
Let's have all our free_and_replace() functions use a single
implementation.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

**Comment by mrc0mmand** [comment]
Hmpf, looking through logs of other jobs, the C8S job might be a bit FUBAR, so I'll delve a bit deeper...

**Comment by medhefgo** [comment]
Why do they return 0 instead of NULL?

**Comment by DaanDeMeyer** [comment]
> Why do they return 0 instead of NULL?

No idea, we don't even use the return value anywhere, let's see if I can get rid of the statement expression altogether.

File: src/basic/alloc-util.h [hunk 1]
```diff
@@ -50,7 +50,7 @@ typedef void* (*mfree_func_t)(void *p);
 
 #define malloc0(n) (calloc(1, (n) ?: 1))
 
-#define free_and_replace(a, b)                  \
+#define free_and_replace_full(a, b, free)       \
```
Comment (yuwata) [inline src/basic/alloc-util.h]: for safety, I'd like to rename `free` -> `free_func` or `_free`.

**Review by yuwata — APPROVED** [review]
LGTM. One minor suggestion.

**Pull request merged**